### ISSUE TITLE
.require with expose works strange

### ISSUE THREAD

**Issue by gobwas** [substack/node-browserify#1131]
Hello! 

Sorry, if it known bug.

This code:

```js
bundler.require("./my/path/to.js", {expose: "./my/path/to.js"})
```

Give the result, when:

```js
// in the bundle

require("./my/path/to.js") // Module not found
require("/my/path/to.js") // Module works O_0
```

Thanks.

**Comment by jmm** [comment]
There's currently a lot of issues with resolution of paths / module IDs. I'm personally not sure that `expose: './whatever'` or `expose: '../whatever'` should be allowed.

```js
require("/my/path/to.js") // Module works O_0
```
Is `/my/path/to.js` the absolute path to the file? Does that work if you omit this?:

```js
bundler.require("./my/path/to.js", {expose: "./my/path/to.js"})
```

**Comment by gobwas** [comment]
No, it is not absolute. It relative in config, but, in the bundle, It like module key, not path.

No, it could not work, cause the situation is something like this:

```js
var config = {path: "./my/path.js"};

setTimeout(function(){ require(config.path); }, 5000);
```

So in the bundle, I think, the `expose` key must be just string key. Nevermind what it really is.

**Comment by jmm** [comment]
> cause the situation is something like this:

That's happening where, in a file that you're bundling?

> So in the bundle, I think, the expose key must be just string key. Nevermind what it really is.

I'm sorry, I couldn't quite follow that -- is there another way you can put it?

**Comment by gobwas** [comment]
Yes, in some file Im bundling.

I mean, I think that `expose` option for `.require` method must just add alias in bundle, not analyze what it is - path or key or smth else.

**Comment by jmm** [comment]
It's not well documented, but with browserify I believe you can only use `require()` calls with literal strings. E.g. I don't believe this will work:

```js
var path = 'whatever';
require(path);
```

Re: the other issue, the problem is like I said that paths and things that look like paths are handled inconsistently in a variety of ways, so `require('./whatever')` might get treated as a relative path instead of an `expose` key.

**Comment by jmm** [comment]
Your example with `setTimeout` is in a file you're bundling or in your build script?

**Comment by gobwas** [comment]
In file that Im bundling. It show, that dependency is not literal string and is not asked synchronously.

**Comment by gobwas** [comment]
I want just to `expose` in `require` add file to bundle and create exposed alias in bundle for it =)

**Comment by jmm** [comment]
> In file that Im bundling. It shows, that dependency is not literal string

I don't believe browserify supports that.

> I just want to add file to bundle and create exposed alias in bundle for it. Then require it somehow and somewhen later. =)

Let's figure out the part about the literal string and see where that leaves us.

**Comment by gobwas** [comment]
Sorry. I did not understand, you don't believe browserify supports what? Browserify does not need to support parsing of not string literals in require calls. It just need to support expected behaviour of `expose` option =)

I mean just one simple little thing:

```js
// inside the build script:

// why this works good:
bundler.require("./my/module.js", { expose: "somehashstring" });

// and why this not:
bundler.require("./my/module.js", { expose: "./my/module.js" });

```

```js
// inside the bundle

// this works alright:
require("somehashstring"); // ok

// but this not!
require("./my/module.js"); // not okay!

// and what is the problem:
require("/my/module.js"); // ok!
```

Do you understand now? )
The problem is that browserify suddenly and wonder why analyzes the path and replaces the first `.` in the string, that passed to `expose` argument.

**Comment by jmm** [comment]
> Sorry. I did not understand, you don't believe browserify supports what?

`require()` calls not with literal strings.

> Browserify does not need to support parsing of not string literals in require calls.

Ok. I didn't understand your example.

> I mean just one simple little thing:
...
> Do you understand now? )

Yes, thanks for clarifying.

As I said, there's currently a lot of issues with resolution of paths / module IDs in browserify. These values that are, or look like, paths are probably being handled in an inconsistent way internally that's causing them to not match up as expected. There also was / is an issue specifically affecting `require()` calls with an `expose` where both values are the same and begin with `./`. E.g. see: https://github.com/substack/module-deps/pull/65#issuecomment-70507321

As I noted, I wonder if `expose` values beginning with `./` or `../` should even be allowed.

I'm actually surprised that in your case `require("/my/module.js");` works, I don't know the explanation for that off the top of my head. I'll try to take a closer look at this issue at some point. In the meantime if you can post a link to a repo that provides a minimal reproduction of the problem, that would be helpful. What version of browserify are you using? (That should always be included in an issue report.)

**Comment by jmm** [comment]
@gobwas If this is still a problem for you, please open a new issue using the guidlines in [How to report issues with browserify](https://github.com/substack/node-browserify/wiki/How-to-report-issues-with-browserify). Thanks!
